Compile probe does not take Cargo rustc configuration into account

The compile probe just invokes `rustc` directly, without including relevant cargo configuration. This causes problems in environments where that configuration may change whether or not backtrace is available (or compiles):

```bash
cargo new anyhow-issue
cd anyhow-issue
mkdir .cargo
echo '[build]' >> .cargo/config.toml
echo 'rustflags = ["-Zallow-features="]' >> .cargo/config.toml
echo 'anyhow = "1"' >> Cargo.toml
rustup override set nightly
cargo check
```

produces

```
    Updating crates.io index
  Downloaded anyhow v1.0.41
  Downloaded 1 crate (34.6 KB) in 0.68s
   Compiling anyhow v1.0.41
error[E0725]: the feature `backtrace` is not in the list of allowed features
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.41/src/lib.rs:213:32
    |
213 | #![cfg_attr(backtrace, feature(backtrace))]
    |                                ^^^^^^^^^

error[E0658]: use of unstable library feature 'backtrace'
 --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.41/src/backtrace.rs:2:33
  |
2 | pub(crate) use std::backtrace::{Backtrace, BacktraceStatus};
  |                                 ^^^^^^^^^
  |
  = note: see issue #53487 <https://github.com/rust-lang/rust/issues/53487> for more information
  = help: add `#![feature(backtrace)]` to the crate attributes to enable
...
```

Taking _everything_ into account is probably too tall of an order, but the [Cargo environment variables](https://doc.rust-lang.org/cargo/reference/environment-variables.html) should get us pretty far. In particular, if set, anyhow should probably take into account these three:

 - `CARGO_BUILD_RUSTC`
 - `CARGO_BUILD_RUSTC_WRAPPER`
 - `CARGO_BUILD_RUSTFLAGS`

Together, those should encapsulate _most_ of the rustc configuration in cargo.

--- a/build.rs
+++ b/build.rs
@@ -67,17 +67,37 @@ fn compile_probe() -> Option<ExitStatus> {
     let out_dir = env::var_os("OUT_DIR")?;
     let probefile = Path::new(&out_dir).join("probe.rs");
     fs::write(&probefile, PROBE).ok()?;
-    Command::new(rustc)
-        .stderr(Stdio::null())
+
+    // Make sure to pick up Cargo rustc configuration.
+    let mut cmd = env::var_os("CARGO_RUSTC_WRAPPER").map_or_else(
+        || Command::new(&rustc),
+        |wrapper| {
+            let mut cmd = Command::new(wrapper);
+            // The wrapper's first argument should always be the path to rustc.
+            cmd.arg(&rustc);
+            cmd
+        },
+    );
+
+    cmd.stderr(Stdio::null())
         .arg("--edition=2018")
         .arg("--crate-name=anyhow_build")
         .arg("--crate-type=lib")
         .arg("--emit=metadata")
         .arg("--out-dir")
         .arg(out_dir)
-        .arg(probefile)
-        .status()
-        .ok()
+        .arg(probefile);
+
+    // If Cargo wants to set RUSTFLAGS, use that.
+    if let Ok(rustflags) = env::var("CARGO_ENCODED_RUSTFLAGS") {
+        if !rustflags.is_empty() {
+            for arg in rustflags.split('\x1f') {
+                cmd.arg(arg);
+            }
+        }
+    }
+
+    cmd.status().ok()
 }
 
 fn rustc_minor_version() -> Option<u32> {

